### DISCUSSION TITLE
End-to-end Learning in Ray-tracing channel

### DISCUSSION THREAD

**Discussion by Mio-TW** [NVlabs/sionna#556] [Q&A]
Dear Sionna Team,

I want to run end-to-end learning in a ray-tracing channel.
So I combined the tutorials from [Autoencoder](https://nvlabs.github.io/sionna/examples/Autoencoder.html) and [Ray-tracing](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Introduction.html), and used the conventional training in my codes (not reinforcement learning case).
The channel is generated using the following codes in [here](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Introduction.html#Site-specifc-Link-Level-Simulations)

```py
# Init CIR generator
cir_generator = CIRGenerator(a,
                             tau,
                             num_tx)

# Initialises a channel model that can be directly used by OFDMChannel layer
channel_model = CIRDataset(cir_generator,
                           batch_size,
                           num_rx,
                           num_rx_ant,
                           num_tx,
                           num_tx_ant,
                           75,
                           num_time_steps)

# Configure the actual channel
self._channel = OFDMChannel(
                     self._channel_model,
                     self._pusch_transmitter.resource_grid,
                     normalize_channel=True,
                     return_channel=True)
```

But I suffer the following error.
It seems that the gradient cannot be backpropagated through the "ofdm channel".

So I have the following questions:
1. Why can't the gradient pass through the OFDM channel?  
Haven't we already calculated the channel response using 'a' and 'tau'? (Equivalent to knowing '**H**' in y=Hx+n) 
Just like passing through an AWGN channel, shouldn't it be enough to pass the gradient through? 
I'm not sure why this error would happen.
Please tell me if my understanding is correct. 
If so, please tell me how to change the code of **OFDM channel** so that it can pass gradient.

2. If I use the "tf.stop_gradient" suggested by the error message, the trainable constellation parameters on the transmitter side will not be updated, which is not the result I want.

Thank you so much.

```
File ~/.local/lib/python3.8/site-packages/tensorflow/python/ops/gradients_util.py:631, in _GradientsHelper(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method, stop_gradients, unconnected_gradients, src_graph)
    629       grad_fn = func_call.python_grad_func
    630     else:
--> 631       raise LookupError(
    632           "No gradient defined for operation"
    633           f"'{op.name}' (op type: {op.type}). "
    634           "In general every operation must have an associated "
    635           "`@tf.RegisterGradient` for correct autodiff, which this "
    636           "op is lacking. If you want to pretend this "
    637           "operation is a constant in your program, you may insert "
    638           "`tf.stop_gradient`. This can be useful to silence the "
    639           "error in cases where you know gradients are not needed, "
    640           "e.g. the forward pass of tf.custom_gradient. "
    641           "Please see more details in "
    642           "https://www.tensorflow.org/api_docs/python/tf/custom_gradient.")  # pylint: disable=line-too-long
    643 if loop_state:
    644   loop_state.EnterGradWhileContext(op, before=False)

No gradient defined for operation'ofdm_channel_1/IteratorGetNext' (op type: IteratorGetNext). 
In general every operation must have an associated `@tf.RegisterGradient` for correct autodiff, which this op is lacking. 
If you want to pretend this operation is a constant in your program, you may insert `tf.stop_gradient`. 
This can be useful to silence the error in cases where you know gradients are not needed, e.g. the forward pass of tf.custom_gradient.
Please see more details in https://www.tensorflow.org/api_docs/python/tf/custom_gradient.
```

**Comment by faycalaa** [comment]
Hi @Mio-TW,

Could you please share a minimal code snippet that reproduces the error?

**Reply by Mio-TW** [reply]
Hi @faycalaa,
Thank you for responding promptly. 
You can copy the provided codes below and run it.
This will trigger the error I mentioned above.

```py
import sionna
import numpy as np
import math
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import Layer, Conv2D, LayerNormalization
from tensorflow.nn import relu

from sionna.channel import OFDMChannel, CIRDataset
from sionna.mimo import StreamManagement
from sionna.ofdm import ResourceGrid, ResourceGridMapper, ResourceGridDemapper
from sionna.fec.ldpc.encoding import LDPC5GEncoder
from sionna.fec.ldpc.decoding import LDPC5GDecoder
from sionna.mapping import Mapper, Demapper, Constellation
from sionna.utils import log10, expand_to_rank
from sionna.utils import BinarySource, ebnodb2no, insert_dims
from sionna.rt import load_scene, Transmitter, Receiver, PlanarArray, Camera, scene

#############################
##      Configuration      ##
#############################
NUM_UT = 1
NUM_BS = 1
NUM_UT_ANT = 1
NUM_BS_ANT = 8

carrier_frequency = 3.5e9   # Hz
subcarrier_spacing = 30e3   # Hz
fft_size = 128
num_ofdm_symbols = 14
cyclic_prefix_length = 10 
num_bits_per_symbol = 6
coderate = 0.5 

batch_size = 32
num_training_iterations = 10000

# SNR range for training [dB]
ebno_db_train_min = -4.0
ebno_db_train_max = 3.0

##########################
##      Ray Tracing     ##
##########################
batch_size_cir = 20
target_num_cirs = 100

scene = load_scene(scene.munich)
scene.frequency = carrier_frequency
resolution = [960,640]

# Create transmitter
scene.remove("tx")
scene.tx_array = PlanarArray(num_rows=2,
                             num_cols=2,
                             vertical_spacing=0.5,
                             horizontal_spacing=0.5,
                             pattern="tr38901",
                             polarization="cross")

tx = Transmitter(name="tx", position=[8.5,21,27])
scene.add(tx)

# Set the receiver array
scene.rx_array = PlanarArray(num_rows=1,
                             num_cols=1,
                             vertical_spacing=0.5,
                             horizontal_spacing=0.5,
                             pattern="iso", # UE orientation is random
                             polarization="V")

# Update coverage_map
cm = scene.coverage_map(max_depth=5,
                        reflection=True,
                        diffraction=True,
                        cm_cell_size=(1., 1.),
                        combining_vec=None,
                        precoding_vec=None,
                        num_samples=int(1e6))

# sample batch_size random user positions from coverage map
ue_pos = cm.sample_positions(batch_size=batch_size_cir,
                             min_gain_db=-130,
                             max_gain_db=0,
                             min_dist=5,
                             max_dist=400)

# Create batch_size receivers
for i in range(batch_size_cir):
    scene.remove(f"rx-{i}")

for i in range(batch_size_cir):
    rx = Receiver(name=f"rx-{i}",
                  position=ue_pos[i],  # Random position sampled from coverage map
                  )
    scene.add(rx)

### Generate the CIRs for many different random positions
# Remark: some path are removed if no path was found for this position
num_runs = int(np.ceil(target_num_cirs / batch_size_cir))
a = None
tau = None
UE_position = []

for idx in range(num_runs):
    print(f"Progress: {idx + 1}/{num_runs}", end="\r")

    # Sample random user positions
    ue_pos = cm.sample_positions(batch_size=batch_size_cir,
                                 min_gain_db=-130,
                                 max_gain_db=0,
                                 min_dist=5,
                                 max_dist=400)

    UE_position.append(ue_pos.numpy())

    # Update all receiver positions
    for idx in range(batch_size_cir):
        scene.receivers[f"rx-{idx}"].position = ue_pos[idx]

    # Simulate CIR
    paths = scene.compute_paths(
        max_depth=5,
        diffraction=True,
        num_samples=1e6)  # shared between all tx in a scene

    # Transform paths into channel impulse responses
    paths.reverse_direction = True  # Convert to uplink direction
    paths.apply_doppler(sampling_frequency=subcarrier_spacing,
                        num_time_steps=num_ofdm_symbols,
                        tx_velocities=[0, 0, 0],
                        rx_velocities=[3.0, 3.0, 0])

    # We fix here the maximum number of paths to 75 which ensures
    # that we can simply concatenate different channel impulse reponses
    a_, tau_ = paths.cir(num_paths=75)

    del paths  # Free memory

    if a is None:
        a = a_.numpy()
        tau = tau_.numpy()
    else:
        # Concatenate along the num_tx dimension
        a = np.concatenate([a, a_], axis=3)
        if scene.synthetic_array == True:
            tau = np.concatenate([tau, tau_], axis=2)
        else:
            tau = np.concatenate([tau, tau_], axis=3)

del cm  # Free memory

# Exchange the num_tx and batchsize dimensions
a = np.transpose(a, [3, 1, 2, 0, 4, 5, 6])
if scene.synthetic_array == True:
    tau = np.transpose(tau, [2, 1, 0, 3])
else:
    tau = np.transpose(tau, [3, 1, 2, 0, 4, 5])

# Remove CIRs that have no active link (i.e., a is all-zero)
p_link = np.sum(np.abs(a) ** 2, axis=(1, 2, 3, 4, 5, 6)) # (5000)
a = a[p_link > 0., ...]
tau = tau[p_link > 0., ...]

UE_position = np.concatenate(UE_position, axis=0)
UE_position = UE_position[p_link > 0., ...]

print("Shape of a:", a.shape)                      
print("Shape of tau: ", tau.shape)                 
print("Shape of UE_position:", UE_position.shape)  

#############################
##      Channel Model      ##
#############################
class CIRGenerator:

    def __init__(self,
                 a,
                 tau,
                 num_tx):

        # Copy to tensorflow
        self._a = tf.constant(a, tf.complex64)
        self._tau = tf.constant(tau, tf.float32)
        self._dataset_size = self._a.shape[0]

        self._num_tx = num_tx

    def __call__(self):

        # Generator implements an infinite loop that yields new random samples
        while True:
            # Sample 4 random users and stack them together
            idx,_,_ = tf.random.uniform_candidate_sampler(
                            tf.expand_dims(tf.range(self._dataset_size, dtype=tf.int64), axis=0),
                            num_true=self._dataset_size,
                            num_sampled=self._num_tx,
                            unique=True,
                            range_max=self._dataset_size)

            a = tf.gather(self._a, idx)
            tau = tf.gather(self._tau, idx)

            # Transpose to remove batch dimension
            a = tf.transpose(a, (3,1,2,0,4,5,6))
            tau = tf.transpose(tau, (2,1,0,3))

            # And remove batch-dimension
            a = tf.squeeze(a, axis=0)
            tau = tf.squeeze(tau, axis=0)

            yield a, tau
            
# Init CIR generator
cir_generator = CIRGenerator(a,
                             tau,
                             num_tx=NUM_UT)

# Initialises a channel model that can be directly used by OFDMChannel layer
channel_model = CIRDataset(cir_generator,
                           batch_size,
                           num_rx=NUM_BS,
                           num_rx_ant=NUM_BS_ANT,
                           num_tx=NUM_UT,
                           num_tx_ant=NUM_UT_ANT,
                           num_paths=75,
                           num_time_steps=num_ofdm_symbols)

# Delete to free memory
del a, tau

##############################
##      NeuralReceiver      ##
##############################
class ResidualBlock(Layer):

    def __init__(self):
        super().__init__()

        # Layer normalization is done over the last three dimensions: time, frequency, conv 'channels'
        self._layer_norm_1 = LayerNormalization(axis=(-1, -2, -3))
        self._conv_1 = Conv2D(filters=128,
                              kernel_size=[3,3],
                              padding='same',
                              activation=None)
        # Layer normalization is done over the last three dimensions: time, frequency, conv 'channels'
        self._layer_norm_2 = LayerNormalization(axis=(-1, -2, -3))
        self._conv_2 = Conv2D(filters=128,
                              kernel_size=[3,3],
                              padding='same',
                              activation=None)

    def call(self, inputs):
        z = self._layer_norm_1(inputs)
        z = relu(z)
        z = self._conv_1(z)
        z = self._layer_norm_2(z)
        z = relu(z)
        z = self._conv_2(z) # [batch size, num time samples, num subcarriers, num_channels]
        # Skip connection
        z = z + inputs

        return z

class Neural_Receiver(Layer):

    def __init__(self):
        super().__init__()

        # Input convolution
        self._input_conv = Conv2D(filters=128,
                                  kernel_size=[3,3],
                                  padding='same',
                                  activation=None)
        # Residual blocks
        self._res_block_1 = ResidualBlock()
        self._res_block_2 = ResidualBlock()
        self._res_block_3 = ResidualBlock()
        self._res_block_4 = ResidualBlock()
        # Output conv
        self._output_conv = Conv2D(filters=num_bits_per_symbol,
                                   kernel_size=[3,3],
                                   padding='same',
                                   activation=None)

    def call(self, inputs):
        y, no = inputs

        # Assuming a single receiver, remove the num_rx dimension
        y = tf.squeeze(y, axis=1)

        # Feeding the noise power in log10 scale helps with the performance
        no = log10(no)

        # Stacking the real and imaginary components of the different antennas along the 'channel' dimension
        y = tf.transpose(y, [0, 2, 3, 1]) # Putting antenna dimension last
        no = insert_dims(no, 3, 1)
        no = tf.tile(no, [1, y.shape[1], y.shape[2], 1])
        # z : [batch size, num ofdm symbols, num subcarriers, 2*num rx antenna + 1]
        z = tf.concat([tf.math.real(y),
                       tf.math.imag(y),
                       no], axis=-1)
        # Input conv
        z = self._input_conv(z)
        # Residual blocks
        z = self._res_block_1(z)
        z = self._res_block_2(z)
        z = self._res_block_3(z)
        z = self._res_block_4(z)
        # Output conv
        z = self._output_conv(z)

        # Reshape the input to fit what the resource grid demapper is expected
        z = insert_dims(z, 2, 1)

        return z
        
        
class E2ESystemConventionalTraining(Model):
    
    def __init__(self, training):
        super().__init__()
        
        self._training = training
        
        # Create an RX-TX association matrix and stream management
        self._num_streams_per_tx = NUM_UT_ANT
        self._rx_tx_association = np.array([[1]])
        self._stream_manager = StreamManagement(self._rx_tx_association, self._num_streams_per_tx)
        
        ############# Configuration #############
        # Resource Grid
        self._resource_grid = ResourceGrid(num_ofdm_symbols=num_ofdm_symbols,
                                           fft_size=fft_size,
                                           subcarrier_spacing=subcarrier_spacing,
                                           num_tx=NUM_UT,  # Uplink
                                           num_streams_per_tx=self._num_streams_per_tx,
                                           cyclic_prefix_length=cyclic_prefix_length,
                                           pilot_pattern="kronecker",
                                           pilot_ofdm_symbol_indices=[2,11])

        # Channel
        self._ofdm_channel = OFDMChannel(channel_model=channel_model,
                                         resource_grid=self._resource_grid,
                                         add_awgn=True,
                                         normalize_channel=True,
                                         return_channel=True)
            
        ############# Transmitter #############
        self._binary_source = BinarySource()
        self._n = int(self._resource_grid.num_data_symbols*num_bits_per_symbol) # Number of coded bits
        self._k = int(self._n*coderate) # Number of information bits
        
        if not self._training:
            self._encoder = LDPC5GEncoder(self._k, self._n, num_bits_per_symbol) 
            
        # Trainable constellation
        self._constellation = Constellation("qam", num_bits_per_symbol, trainable=True)
        self._mapper = Mapper(constellation=self._constellation)
        self._rg_mapper = ResourceGridMapper(self._resource_grid)

        
        ############# Receiver #############
        self._rg_demapper = ResourceGridDemapper(self._resource_grid,
                                                 self._stream_manager)
        self._neural_receiver = Neural_Receiver()

        # To reduce the computational complexity of training, the outer code is not used when training,
        # as it is not required
        if not self._training:
            self._decoder = LDPC5GDecoder(self._encoder, hard_out=True)
            
        ############# Loss Function #############
        if self._training:
            self._bce = tf.keras.losses.BinaryCrossentropy(from_logits=True)

    
    @tf.function(jit_compile=True)
    def call(self, batch_size, ebno_db):
        
        # If `ebno_db` is a scalar, a tensor with shape [batch size] is created as it is what is expected by some layers
        if len(ebno_db.shape) == 0:
            ebno_db = tf.fill([batch_size], ebno_db)
        
        ############# Transmitter #############
        # Outer coding is only performed if not training
        if self._training:
            c = self._binary_source([batch_size, NUM_UT, self._num_streams_per_tx, self._n])
        else:
            b = self._binary_source([batch_size, NUM_UT, self._num_streams_per_tx, self._k])
            c = self._encoder(b)
        
        x = self._mapper(c) 
        x_rg = self._rg_mapper(x)
        
        ############# Channel #############
        no = ebnodb2no(ebno_db, num_bits_per_symbol, coderate)
        no_ = expand_to_rank(no, tf.rank(x_rg))
        y, h = self._ofdm_channel([x_rg, no_])

        ############# Receiver #############
        llr = self._neural_receiver([y, no])
        llr = self._rg_demapper(llr) 
        llr = tf.reshape(llr, [batch_size, NUM_UT, self._num_streams_per_tx, self._n]) 
        if self._training:
            loss = self._bce(c, llr)
            return loss
        else:
            bits_hat = self._decoder(llr)
            return bits, bits_hat      
            
########################
##      Training      ##
########################
# Set a seed for reproducibility
tf.random.set_seed(1)

# Instantiating the end-to-end model for training
model = E2ESystemConventionalTraining(training=True)

# Adam optimizer (SGD variant)
optimizer = tf.keras.optimizers.Adam()

# Training loop
for i in range(num_training_iterations):
    # Sample a batch of SNRs.
    ebno_db = tf.random.uniform(shape=[batch_size], minval=ebno_db_train_min, maxval=ebno_db_train_max)
    
    # Forward pass
    with tf.GradientTape() as tape:
        loss = model(batch_size, ebno_db)
        
    # Computing and applying gradients
    weights = model.trainable_weights
    grads = tape.gradient(loss, weights)
    optimizer.apply_gradients(zip(grads, weights))
    
    # Print progress
    if i % 10 == 0:
        print(f"{i}/{num_training_iterations}  Loss: {loss:.2E}", end="\r")    

```

**Reply by Mio-TW** [reply]
Hi @faycalaa, 
please kindly help check it. 
Thank you so much.

**Reply by jhoydis** [reply]
Hi @Mio-TW,

Have you tried deactivating XLA, i.e., `@tf.function(jit_compile=False)`, or running the training in Eager mode, i.e., remove the `tf.function` decorator completely? I think that the CIR generator does not work in XLA.

**Reply by Mio-TW** [reply]
Hi @jhoydis,

I added the command `sionna.config.xla_compat=False` at the beginning of the code.
Also the `@tf.function(jit_compile=False)` before the calling function of `CIRGenerator`.
But still got the same error.

**Reply by jhoydis** [reply]
Did you change the decorator of the `call` method of the class `E2ESystemConventionalTraining`?

**Reply by Mio-TW** [reply]
Hi @jhoydis, 
Yes I also changed it into `False`.
